samples/app-cache-meetings/nodejs/src/components/app-cache-tab.tsx

```diff
@@ -0,0 +1,126 @@
+// <copyright file="app-cache-tab.tsx" company="Microsoft Corporation">
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+// </copyright>
+
+import { useState } from "react";
+import React from "react";
+import * as microsoftTeams from "@microsoft/teams-js";
+
+/// </summary>
+/// function logitem
+/// </summary>
+function logItem(action: string, actionColor: string, message: string) {
+  const newItem =
+    "<span style='font-weight:bold;color:" +
+    actionColor +
+    "'>" +
+    action +
+    "</span> " +
+    message +
+    "</br>";
+  return newItem;
+}
+
+/// </summary>
+/// In beforeUnloadHandler using setItems and readyToUnload callback function
+/// </summary>
+const beforeUnloadHandler = (
+  setItems: React.Dispatch<React.SetStateAction<string[]>>,
+  readyToUnload: () => void) => {
+
+  let newItem = logItem("OnBeforeUnload", "purple", "Started");
+  setItems((Items) => [...Items, newItem]);
+
+  newItem = logItem("OnBeforeUnload", "purple", "Completed");
+  setItems((Items) => [...Items, newItem]);
+
+  console.log("sending readyToUnload to TEAMS");
+  readyToUnload();
+
+  return true;
+};
+
+/// </summary>
+/// loadHandler 
+/// </summary>
+const loadHandler = (
+  setItems: React.Dispatch<React.SetStateAction<string[]>>,
+  data: microsoftTeams.LoadContext) => {
+  logItem("OnLoad", "blue", "Started for " + data.entityId);
+
+  let newItem = logItem("OnLoad", "blue", "Completed for " + data.entityId);
+  setItems((Items) => [...Items, newItem]);
+
+  microsoftTeams.app.notifySuccess();
+};
+
+const AppCacheTab = () => {
+  const [items, setItems] = useState<string[]>([]);
+  const [title, setTitle] = useState("App Cache Testing Sample");
+  const [initState] = useState(true);
+
+  React.useEffect(() => {
+    if (!initState) {
+      return;
+    }
+
+    // get context
+    microsoftTeams.app.initialize().then(() => {
+      microsoftTeams.app.getContext().then((context) => {
+
+        const newItem = logItem("Success", "green", "Loaded Teams context");
+        setItems((Items) => [...Items, newItem]);
+        setTitle(title);
+
+        const newLogItem = logItem("FrameContext", "orange", "Frame context is " + context.page.frameContext);
+        setItems((Items) => [...Items, newLogItem]);
+
+        if (context.page.frameContext === "sidePanel") {
+          // OnBeforeUnload
+          microsoftTeams.teamsCore.registerBeforeUnloadHandler((readyToUnload) => {
+            const result = beforeUnloadHandler(setItems, readyToUnload);
```

Comment:
Can you add comments to explain what is actually happening here? What does calling the load/unload handlers have to do with caching the items? How does using React State lead to the values being cached across reloads?

Reply:
Yeah, as we know that according to the doc this specifically Apps loaded in the Side Panel of a Meeting.
Steps :
1.Initializing App using **microsoftTeams.app.initialize().**
2.Get context **microsoftTeams.app.getContext()** , when this app loaded in sidepanel of a meeting then the value of this key **framecontext** will be **sidePanel**.
3.Register load and before unload handlers using this method **microsoftTeams.registerBeforeUnloadHandler and microsoftTeams.registerOnLoadHandler** and ready for app caching.
4.The useState hook is returning an array with two items the other item can set the value or in other words change the state.
5.The user of the function can set the state directly or use the function with the previous state as an argument.